### PR TITLE
Update hashbrown requirement from 0.14.2 to 0.15.0

### PR DESCRIPTION
**Pull request by wiedld**
# Which issue does this PR close?

Alternative PR to https://github.com/apache/arrow-rs/pull/6486.

# Rationale for this change
 
Update the ver 0.15 of hashbrown.

We are not using the deprecated `RawTable`, only raw entries.
In order to continue using raw entries, we need to use the [inline-more](https://docs.rs/hashbrown/0.15.0/src/hashbrown/raw_entry.rs.html#106) and `raw-entry` feature flags.

# What changes are included in this PR?

See above.

# Are there any user-facing changes?

No

### CODE REVIEW

File: arrow-flight/gen/src/main.rs [hunk 1]
```diff
@@ -29,7 +29,7 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
         // protoc in unbuntu builder needs this option
         .protoc_arg("--experimental_allow_proto3_optional")
         .out_dir("src")
-        .compile_with_config(prost_config(), &[proto_path], &[proto_dir])?;
```
Comment (wiedld) [inline arrow-flight/gen/src/main.rs:32]: This is a tonic deprecation warning, fixed here. I can move to another PR if a problem.

**Comment by crepererum** [comment]
From the change log:

> Raw entry API is now under raw-entry feature, to be eventually removed.

So if we don't fix that during the upgrade, we should track the migration in a ticket so we're not getting into a pickle when that functionality is actually removed.

**Comment by Jefffrey** [comment]
> From the change log:
> 
> > Raw entry API is now under raw-entry feature, to be eventually removed.
> 
> So if we don't fix that during the upgrade, we should track the migration in a ticket so we're not getting into a pickle when that functionality is actually removed.

Raised https://github.com/apache/arrow-rs/issues/6498

File: arrow-array/Cargo.toml [hunk 2]
```diff
@@ -48,7 +48,7 @@ chrono = { workspace = true }
 chrono-tz = { version = "0.10", optional = true }
 num = { version = "0.4.1", default-features = false, features = ["std"] }
 half = { version = "2.1", default-features = false, features = ["num-traits"] }
-hashbrown = { version = "0.14.2", default-features = false }
+hashbrown = { version = "0.15.0", default-features = false, features = ["inline-more", "raw-entry"] }
```
Comment (Jefffrey) [inline arrow-array/Cargo.toml:51]: Is setting `inline-more` strictly necessary? I think it's mutually exclusive from `raw-entry` as can compile without needing it :thinking:

**Comment by alamb** [comment]
Before we merge this, we should certainly run benchmarks given how important hashbrown is to performance

ALso there is a bunch of discussion downstream in DataFusion (which uses the lower level RawEntry APIs) on what to do now that hashbrown plans to remove it

File: arrow-array/Cargo.toml (hunk 2 above)
Comment (wiedld) [reply arrow-array/Cargo.toml:51]: I got that from reading [the src](https://docs.rs/hashbrown/0.15.0/src/hashbrown/raw_entry.rs.html#106). It's [enabled by default](https://github.com/rust-lang/hashbrown/blob/e057e87ee818df3855b9c8a4ae281cd497479311/README.md?plain=1#L63) for better performance, but we don't compile with the default features. Altho this points out how I should have really run the performance test. Thank you!

I'm also working on an alternative approach to remove use of raw-entry. I hope to performance test both approaches later today.

**Comment by alamb** [comment]
I also wanted to point out given how fundamental hashbrown is to a bunch of things (esp in DataFusion) I think we should let the new `0.15.0` release have some bake time (aka let other projects on the bleeding edge test it and find any bugs) 

Thus I think we should target upgrading in the next arrow release (0.52.2) not the one I am preparing now

**Comment by wiedld** [comment]
> Thus I think we should target upgrading in the next arrow release (0.52.2) not the one I am preparing now

Understood. I'll circle back to this one (or anyone else can feel free to take a shot), after I address some higher priority items. I'm partially into an alternative solution, but it's not near ready.

I'll comment here when I pick it up again, if someone else doesn't beat me to it. 😆

**Comment by tustvold** [comment]
FWIW I've filed an upstream ticket related to this, to see if the maintainers have any recommendations on how we should move forward here - https://github.com/rust-lang/hashbrown/issues/573

**Comment by crepererum** [comment]
With #6537 merged, I think we no longer need the `raw-entry` feature.

**Comment by crepererum** [comment]
Covered by #6684.

**Pull request closed without merging**